Add v22.1.7 release notes (#15107)
cockroachdb/docs: _includes/releases/v22.1/v22.1.7.md

## v22.1.7

Release Date: September 15, 2022

{% include releases/release-downloads-docker-image.md release=include.release %}

<h3 id="v22-1-7-enterprise-edition-changes">Enterprise edition changes</h3>

> - The new `kv.rangefeed.range_stuck_threshold` (default 0, i.e., disabled) [cluster setting](../v22.1/cluster-settings.html) instructs rangefeed clients (used internally by [changefeeds](../v22.1/create-and-configure-changefeeds.html)) to restart automatically if no checkpoint or other event has been received from the server for some time. This is a defense-in-depth mechanism which will log output as follows if triggered: `restarting stuck rangefeed: waiting for r100 (n1,s1):1 [threshold 1m]: rangefeed restarting due to inactivity`. [#87253][#87253]

Author: Fixed

<h3 id="v22-1-7-sql-language-changes">SQL language changes</h3>

> - Added a new [cluster setting](../v22.1/cluster-settings.html) `sql.stats.response.show_internal` with default value of `false` can be set to `true` to display information about internal stats on the [SQL Activity page](../v22.1/ui-sql-dashboard.html), with fingerprint option. [#86869][#86869]

Author: Fixed

> - [`EXPLAIN ANALYZE`](../v22.1/explain-analyze.html) output now contains a warning when the estimated row count for scans is inaccurate and includes a hint to collect the table [statistics](../v22.1/cost-based-optimizer.html#table-statistics) manually. [#86871][#86871]

Author: Fixed

> - Allowed mismatched type numbers in `PREPARE` statements. [#87161][#87161]

Author: Fixed

- Decreased the cardinality of the number on `__moreN__` when replacing literals. [#87269][#87269]
- The structured payloads used for telemetry logs now include the new `Regions` field which indicates the [regions](../v22.1/multiregion-overview.html#database-regions) of the nodes where SQL processing ran for the query. [#87466][#87466]
- Added the schema name to [index](../v22.1/indexes.html) usage statistics telemetry. [#87624][#87624]
- Added a creation timestamp to [index](../v22.1/indexes.html) usage statistics telemetry. [#87624][#87624]

<h3 id="v22-1-7-command-line-changes">Command-line changes</h3>

> - The `\c` metacommand in the [SQL client](../v22.1/cockroach-sql.html) no longer shows the password in plaintext. [#87548][#87548]

Author: Fixed

<h3 id="v22-1-7-db-console-changes">DB Console changes</h3>

> - Use plan gist instead of plan ID on plans table inside the *Explain Plan* tab of the [Statement Details](../v22.1/ui-statements-page.html) page. Also added the plan gist as the first line on the actual *Explain Plan* display. [#86872][#86872]

Author: Fixed

> - Added new 'Last Execution Time' to Statements, which is hidden by default. [#87077][#87077]

Reviewer: ```suggestion
- Added a **Last Execution Time** column to the statements table on the [**Statements** page](../v22.1/ui-statements-page.html). This column is hidden by default. [#87077][#87077]
```

Author: Fixed

> - Added Transaction and Statement fingerprint IDs to correlating [SQL Activity](../v22.1/ui-sql-dashboard.html) overview pages. New columns are hidden by default. [#87100][#87100]

Reviewer: ```suggestion
- Added **Transaction Fingerprint ID** and **Statement Fingerprint ID** columns to the corresponding [**SQL Activity**](../v22.1/ui-sql-dashboard.html) overview pages. These columns are hidden by default. [#87100][#87100]
```

Author: Fixed

> - Properly formatted the execution count under the [Statement Details](../v22.1/ui-statements-page.html) page. Increased the timeout for the Statement Details page so it shows a proper timeout error when it happens, no longer crashing the page. [#87209][#87209]

Reviewer: ```suggestion
- Properly formatted the **Execution Count** under the [**Statement Fingerprints**](../v22.1/ui-statements-page.html#statement-fingerprint-page) page. Increased the timeout for the **Statement Fingerprints** page so it shows a proper timeout error when it happens, no longer crashing the page. [#87209][#87209]
```

Author: Fixed

<h3 id="v22-1-7-bug-fixes">Bug fixes</h3>

- Fixed a vulnerability in the [optimizer](../v22.1/cost-based-optimizer.html) that could cause a panic in rare cases when planning complex queries with [`ORDER BY`](../v22.1/order-by.html). [#86804][#86804]

> - Prior to this fix, queries with many [joins](../v22.1/joins.html) and projections of multicolumn expressions, e.g. `col1 + col2`, either present in the query or within a [virtual column](../v22.1/computed-columns.html) definition, could experience very long optimization times or hangs, where the query is never sent for execution. This commit adds the `disable_hoist_projection_in_join_limitation` [session flag](../v22.1/set-vars.html#supported-variables). [#85871][#85871]

Reviewer: ```suggestion
- Previously, queries with many [joins](../v22.1/joins.html) and projections of multicolumn expressions (e.g., `col1 + col2`), either present in the query or within a [virtual column](../v22.1/computed-columns.html) definition, could experience very long optimization times or hangs, where the query is never sent for execution. This is now fixed by adding the `disable_hoist_projection_in_join_limitation` [session flag](../v22.1/set-vars.html#supported-variables). [#85871][#85871]
```

Author: Fixed

- Fixed a crash/panic that could occur if placeholder arguments were used with the `with_min_timestamp` or `with_max_staleness` [functions](../v22.1/functions-and-operators.html). [#86881][#86881]
- Fixed a crash that could happen when formatting queries that have placeholder `BitArray` arguments. [#86885][#86885]
- CockroachDB now more precisely respects the `distsql_workmem` [setting](../v22.1/set-vars.html#supported-variables) which improves the stability of each node and makes OOMs less likely. [#86916][#86916]
- Previously, escaping a double quote (`"`) with [`COPY`](../v22.1/copy-from.html) in `CSV` mode could ignore all subsequent lines in the same `COPY` if an `ESCAPE` clause were specified. This is now resolved. [#86977][#86977]
- Fixed a bug that caused some special characters to be misread if they were being read by [`COPY ... FROM`](../v22.1/copy-from.html) into a `TEXT[]` column. [#86887][#86887]

> - Timescale object is now properly constructed from session storage, preventing  bugs and crashes in pages that use the timescale object when reloading the page [#86975][#86975]

Reviewer: ```suggestion
- Timescale object is now properly constructed from session storage, preventing bugs and crashes in pages that use the timescale object when reloading the page. [#86975][#86975]
```

Author: Fixed

> - Previously, CockroachDB would return an internal error when evaluating the `json_build_object` [builtin](../v22.1/functions-and-operators.html) when an [enum](../v22.1/enum.html) or a void datum were passed as the first argument, and this is now fixed. [#86851][#86851]

Reviewer: ```suggestion
- Previously, CockroachDB would return an internal error when evaluating the `json_build_object` [builtin](../v22.1/functions-and-operators.html) when an [enum](../v22.1/enum.html) or a void datum were passed as the first argument. This is now fixed. [#86851][#86851]
```

Author: Fixed

- The statement tag for the [`SHOW`](../v22.1/show-vars.html) command results in the pgwire protocol no longer containing the number of returned rows. [#87126][#87126]
- Fixed a bug where the options given to the [`BEGIN TRANSACTION`](../v22.1/begin-transaction.html) command would be ignored if the `BEGIN` was a prepared statement. [#87126][#87126]

> - A bug has been fixed that caused internal errors like "unable to [vectorize](../v22.1/vectorized-execution.html) execution plan: unhandled expression type" in rare cases. [#87182][#87182]

Reviewer: ```suggestion
- Fixed a bug that caused internal errors like "unable to [vectorize](../v22.1/vectorized-execution.html) execution plan: unhandled expression type" in rare cases. [#87182][#87182]
```

Author: Done!

> - The **Explain Tab** inside the [Statement Details](../v22.1/ui-statements-page.html) page now groups plans that have the same shape but a different number of spans in corresponding scans. [#87211][#87211]

Reviewer: ```suggestion
- The **Explain Plans** tab inside the [**Statement Fingerprint**](../v22.1/ui-statements-page.html#statement-fingerprint-page) page now groups plans that have the same shape but a different number of spans in corresponding scans. [#87211][#87211]
```

Author: Updated!

- A bug in the column backfiller, which is used to add or remove columns from tables, failed to account for the need to read [virtual columns](../v22.1/computed-columns.html) which were part of a [primary key](../v22.1/primary-key.html). [Hash-sharded](../v22.1/hash-sharded-indexes.html) [indexes](../v22.1/indexes.html), starting in v22.1, use [virtual columns](../v22.1/computed-columns.html). Any [hash-sharded](../v22.1/hash-sharded-indexes.html) table created in v22.1 or any table created with a virtual column as part of its primary key would indefinitely fail to complete a [schema change](../v22.1/online-schema-changes.html) which adds or removes columns. This bug has been fixed. [#87272][#87272]
- Added a missing memory accounting call when appending a KV to the underlying `kvBuf`. [#87118][#87118]

> - Some [upgrade](../v22.1/upgrade-cockroach-version.html) migrations perform [schema changes](../v22.1/online-schema-changes.html) on system tables. Those upgrades which added [indexes](../v22.1/indexes.html) could, in some cases, get caught retrying because they failed to detect that the migration had already occurred due to the existence of a populated field. When that happens, the finalization of the new version can hang indefinitely and require manual intervention. This bug has been fixed. [#87633][#87633]

Reviewer: ```suggestion
- Some [upgrade](../v22.1/upgrade-cockroach-version.html) migrations perform [schema changes](../v22.1/online-schema-changes.html) on system tables. Those upgrades which added [indexes](../v22.1/indexes.html) could, in some cases, get caught retrying because they failed to detect that the migration had already occurred due to the existence of a populated field. When that happens, the finalization of the new version could hang indefinitely and require manual intervention. This bug has been fixed. [#87633][#87633]
```

Author: Fixed!

- Fixed a bug that led to the `querySummary` field in the `crdb_internal.statements_statistics`' metadata column being empty. [#87618][#87618]
- Previously, the `querySummary` metadata field in the `crdb_internal.statement_statistics` table was inconsistent with the query metadata field for executed prepared statements. These fields are now consistent for prepared statements. [#87618][#87618]
- Fixed a rare bug where errors could occur related to the use of [arrays](../v22.1/array.html) of [enums](../v22.1/enum.html). [#85961][#85961]
- Fixed a bug that would result in a failed cluster [restore](../v22.1/restore.html). [#87764][#87764]

> - This patch fixes a misused query optimization involving tables with one or more [`PARTITION BY`](../v22.1/partition-by.html) clauses and partition [zone constraints](../v22.1/configure-replication-zones.html) which assign [region locality](../v22.1/set-locality.html) to those partitions. In some cases the [optimizer](../v22.1/cost-based-optimizer.html) picks a ['locality-optimized search'](../v22.1/cost-based-optimizer.html#locality-optimized-search-in-multi-region-clusters) query plan which is not truly locality-optimized, and has higher latency than competing query plans which use distributed scan. Locality-optimized search is now avoided in cases which are known not to benefit from this optimization. [#87848][#87848]

Reviewer: ```suggestion
- Fixed a misused query optimization involving tables with one or more [`PARTITION BY`](../v22.1/partition-by.html) clauses and partition [zone constraints](../v22.1/configure-replication-zones.html) which assign [region locality](../v22.1/set-locality.html) to those partitions. In some cases the [optimizer](../v22.1/cost-based-optimizer.html) picks a ['locality-optimized search'](../v22.1/cost-based-optimizer.html#locality-optimized-search-in-multi-region-clusters) query plan which is not truly locality-optimized, and has higher latency than competing query plans which use distributed scan. Locality-optimized search is now avoided in cases which are known not to benefit from this optimization. [#87848][#87848]
```

Author: Fixed

<h3 id="v22-1-7-performance-improvements">Performance improvements</h3>

- Planning time has been reduced for queries over tables with a large number of columns and/or [indexes](../v22.1/indexes.html). [#86749][#86749]
- Long-running SQL sessions are now less likely to maintain large allocations for long periods of time, which decreases the risk of OOM and improves memory utilization. [#86797][#86797]

<h3 id="v22-1-7-build-changes">Build changes</h3>

> - Fixed OSS builds that didn't have CCL-licensed UI intermediates lingering on-disk. [#86425][#86425]

Reviewer: ```suggestion
- Fixed OSS builds that did not have CCL-licensed UI intermediates lingering on-disk. [#86425][#86425]
```

Author: Updated
